File: Aliucord/src/main/java/com/aliucord/api/PatcherAPI.java
/*
 * Copyright (c) 2021 Juby210 & Vendicated
 * Licensed under the Open Software License version 3.0
 */

package com.aliucord.api;

import androidx.annotation.NonNull;

import com.aliucord.patcher.*;

import java.lang.reflect.Member;
import java.util.ArrayList;
import java.util.List;

import de.robv.android.xposed.XC_MethodHook;
import rx.functions.Action1;
import top.canyie.pine.callback.MethodHook;

@SuppressWarnings({"unused", "deprecation"})
public class PatcherAPI {
    public List<Runnable> unpatches = new ArrayList<>();

    private Runnable createUnpatch(Runnable _unpatch) {
        Runnable unpatch = new Runnable() {
            public void run() {
                _unpatch.run();
                unpatches.remove(this);
            }
        };
        unpatches.add(unpatch);
        return unpatch;
    }

    private Runnable createUnpatch(XC_MethodHook.Unhook unhook) {
        Runnable unpatch = new Runnable() {
            public void run() {
                unhook.unhook();
                unpatches.remove(this);
            }
        };
        unpatches.add(unpatch);
        return unpatch;
    }


    /**
     * Patches a method.
     *
     * @param forClass   Class to patch.
     * @param fn         Method to patch.
     * @param paramTypes Parameters of the <code>fn</code>. Useful for patching individual overloads.
     * @param hook       Callback for the patch.
     * @return A {@link Runnable} object.
     * @see PinePatchFn
     * @see PinePrePatchFn
     * @deprecated Use {@link #patch(String, String, Class[], XC_MethodHook)}
     */
    @Deprecated
    public Runnable patch(@NonNull String forClass, @NonNull String fn, @NonNull Class<?>[] paramTypes, @NonNull MethodHook hook) {
        return createUnpatch(Patcher.addPatch(forClass, fn, paramTypes, hook));
    }

    /**
     * Patches a method.
     *
     * @param clazz      Class to patch.
     * @param fn         Method to patch.
     * @param paramTypes Parameters of the <code>fn</code>. Useful for patching individual overloads.
     * @param hook       Callback for the patch.
     * @return Method that will remove the patch when invoked
     * @see PinePatchFn
     * @see PinePrePatchFn
     * @deprecated Use {@link #patch(Class, String, Class[], XC_MethodHook)}
     */
    @Deprecated
    public Runnable patch(@NonNull Class<?> clazz, @NonNull String fn, @NonNull Class<?>[] paramTypes, @NonNull MethodHook hook) {
        return createUnpatch(Patcher.addPatch(clazz, fn, paramTypes, hook));
    }

    /**
     * Patches a method or constructor.
     *
     * @param m    Method or constructor to patch. see {@link Member}.
     * @param hook Callback for the patch.
     * @return Method that will remove the patch when invoked
     * @see PatcherAPI#patch(String, String, Class[], MethodHook)
     * @see PatcherAPI#patch(Class, String, Class[], MethodHook)
     * @see PinePatchFn
     * @see PinePrePatchFn
     * @deprecated Use {@link #patch(Member, XC_MethodHook)}
     */
    @Deprecated
    public Runnable patch(@NonNull Member m, @NonNull MethodHook hook) {
        return createUnpatch(Patcher.addPatch(m, hook));
    }


    /**
     * Patches a method.
     *
     * @param forClass   Class to patch.
     * @param fn         Method to patch.
     * @param paramTypes Parameters of the <code>fn</code>. Useful for patching individual overloads.
     * @param hook       Callback for the patch.
     * @return A {@link Runnable} object.
     */
    public Runnable patch(@NonNull String forClass, @NonNull String fn, @NonNull Class<?>[] paramTypes, @NonNull XC_MethodHook hook) {
        return createUnpatch(Patcher.addPatch(forClass, fn, paramTypes, hook));
    }

    /**
     * Patches a method.
     *
     * @param clazz      Class to patch.
     * @param fn         Method to patch.
     * @param paramTypes Parameters of the <code>fn</code>. Useful for patching individual overloads.
     * @param hook       Callback for the patch.
     * @return Method that will remove the patch when invoked
     */
    public Runnable patch(@NonNull Class<?> clazz, @NonNull String fn, @NonNull Class<?>[] paramTypes, @NonNull XC_MethodHook hook) {
        return createUnpatch(Patcher.addPatch(clazz, fn, paramTypes, hook));
    }

    /**
     * Patches a method or constructor.
     *
     * @param m    Method or constructor to patch. see {@link Member}.
     * @param hook Callback for the patch.
     * @return Method that will remove the patch when invoked
     */
    public Runnable patch(@NonNull Member m, @NonNull XC_MethodHook hook) {
        return createUnpatch(Patcher.addPatch(m, hook));
    }

    /**
     * Patches a method or constructor.
     *
     * @param m    Method or constructor to patch. see {@link Member}.
     * @param callback Callback for the patch.
     * @return Method that will remove the patch when invoked
     */
    public Runnable patch(@NonNull Member m, @NonNull Action1<XC_MethodHook.MethodHookParam> callback) {
        return createUnpatch(Patcher.addPatch(m, new Hook(callback)));
    }

    /**
     * Removes all patches.
     */
    public void unpatchAll() {
        Object[] runnables = unpatches.toArray();
        for (Object unpatch : runnables) ((Runnable) unpatch).run();
    }
}

File: Aliucord/src/main/java/com/aliucord/api/PatcherAPI.kt
/*
 * Copyright (c) 2021 Juby210 & Vendicated
 * Licensed under the Open Software License version 3.0
 */
package com.aliucord.api

import com.aliucord.Logger
import com.aliucord.patcher.Hook
import com.aliucord.patcher.MethodHookCallback
import com.aliucord.patcher.Patcher.addPatch
import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.XC_MethodHook.Unhook
import java.lang.reflect.Member

typealias Unpatch = Runnable

/**
 * Runtime patching helper that wraps Xposed hooks and tracks created unpatch actions.
 *
 * Use instances to apply hooks via the `patch` methods. Each patch returns an [Unpatch]
 * callback that will remove the applied hook when invoked. Call [unpatchAll] to
 * run all registered unpatch actions and clear the internal registry.
 *
 * @param logger logger used for error reporting
 */
class PatcherAPI internal constructor(
    @PublishedApi
    internal val logger: Logger
) {
    private val unpatches: MutableList<Unpatch> = mutableListOf()

    private fun createUnpatch(unhook: Unhook?): Unpatch {
        val unpatch = Unpatch { unhook?.unhook() }
        unpatches += unpatch
        return unpatch
    }

    /**
     * Patch a method by class name.
     *
     * @param forClass full name of the class containing the method
     * @param fn method name to patch
     * @param paramTypes parameter types to select an overload
     * @param hook the [XC_MethodHook] callback to apply
     * @return an [Unpatch] that will remove the applied hook
     */
    @JvmOverloads
    fun patch(
        forClass: String,
        fn: String,
        paramTypes: Array<Class<*>> = emptyArray(),
        hook: XC_MethodHook
    ): Unpatch {
        return createUnpatch(addPatch(forClass, fn, paramTypes, hook))
    }

    /**
     * Patch a method by [Class].
     *
     * @param clazz the class containing the method
     * @param fn method name to patch
     * @param paramTypes parameter types to select an overload
     * @param hook the [XC_MethodHook] callback to apply
     * @return an [Unpatch] that will remove the applied hook
     */
    @JvmOverloads
    fun patch(
        clazz: Class<*>,
        fn: String,
        paramTypes: Array<Class<*>> = emptyArray(),
        hook: XC_MethodHook
    ): Unpatch {
        return createUnpatch(addPatch(clazz, fn, paramTypes, hook))
    }

    /**
     * Patch a specific method or constructor.
     *
     * @param m the reflective [Member] (method or constructor) to hook
     * @param hook the [XC_MethodHook] callback to apply
     * @return an [Unpatch] that will remove the applied hook
     */
    fun patch(m: Member, hook: XC_MethodHook): Unpatch {
        return createUnpatch(addPatch(m, hook))
    }

    /**
     * Patch a specific method or constructor using a [MethodHookCallback].
     *
     * @param m the reflective [Member] (method or constructor) to hook
     * @param callback the callback wrapper to apply
     * @return an [Unpatch] that will remove the applied hook
     */
    fun patch(m: Member, callback: MethodHookCallback): Unpatch {
        return createUnpatch(addPatch(m, Hook(callback)))
    }

    /**
     * Invoke all registered unpatch actions and clear the registry.
     *
     * This will remove all applied hooks that were created through this [PatcherAPI] instance.
     */
    fun unpatchAll() {
        unpatches.forEach(Unpatch::run)
        unpatches.clear()
    }
}
